### DISCUSSION TITLE
Question about VSG units

### DISCUSSION THREAD

**Discussion by xiaoheimai95** [CURENT/andes#439] [Q&A] [answered]
Dear Dr. Cui and Jinning,
We would like to know whether VSG model (REGCV1) in our platform takes the threshold of current limiter into consideration. And if yes, can you give me some materials about that?

**Comment by jinningwang** [comment, marked as answer]
After checking the [code](https://github.com/CURENT/andes/blob/master/andes/models/renewable/regcv1.py) and [documentation](https://docs.andes.app/en/latest/groupdoc/RenGen.html#regcv1), I don't think there is any ``Limiter`` in ``REGCV1``.

I am happy to have a further discussion if you want to implement any practical or reasonable ``Limiter``.

In the future, you can use the same procedures to check the detailed implementation of any model in ANDES. This is one of the advantages of ANDES over black-box software, as it allows you to know the details of the model and make any necessary modifications.

**Reply by xiaoheimai95** [reply]
That's so nice of you, Jinning! I have referred to some materails, and the current limitation mainly lies in `Idref` and `Iqref`, which will affect the output of `Pe.` Can you suggest where in the code I should make some changes?

**Reply by jinningwang** [reply]
There are some pre-defined blocks [andes.core.block.PIController](https://docs.andes.app/en/latest/_modules/andes/core/block.html?highlight=PIController#) that you might use. For example, You can replace ``PIController`` with ``PIAWHardLimit`` in the definition of [VSGOuterPIModel](https://github.com/CURENT/andes/blob/1f619d0527972356f1c304063036b1f06c884b67/andes/models/renewable/regcv1.py#L282) as below. Here I skipped the detailed parameters input, you might need to pay attention.

```python
class VSGOuterPIModel:
    """
    Outer PI controllers for REGCV1
    """

    def __init__(self, vderr: str = 'vd-vref2', vqerr: str = 'vq'):
        self.PIvd = PIAWHardLimit(u=vderr,
                                 kp=self.Kpvd,
                                 ki=self.Kivd,
                                 x0='Id0',
                                 )
        self.PIvq = PIAWHardLimit(u=vqerr,
                                 kp=self.Kpvq,
                                 ki=self.Kivq,
                                 x0='Iq0',
                                 )

        self.Idref = AliasAlgeb(self.PIvd_y)
        self.Iqref = AliasAlgeb(self.PIvq_y)
```

I would suggest referencing some industry-grade models and following their design. [WECC Approved Dynamic Model Library](https://www.wecc.org/Reliability/Approved%20Dynamic%20Models%20May%202022.pdf) can be a good reference. Further, [Examples](https://docs.andes.app/en/latest/modeling/examples.html) and [Model development workflow](https://github.com/CURENT/andes/discussions/241) might be helpful for model development.

**Reply by xiaoheimai95** [reply]
Thank you for your answer! 
The equation of `Idref`, `Iqref` and `Pe` under current limitation are as below:
<img width="568" alt="截屏2023-04-14 21 04 27" src="https://user-images.githubusercontent.com/71546641/232051470-c887aafb-25eb-4d4c-91db-0cf8adfa812f.png">
<img width="337" alt="截屏2023-04-14 21 04 53" src="https://user-images.githubusercontent.com/71546641/232051552-43c9be14-cbc5-45d6-b138-d18b53bf3a03.png">
So actually, I would like to add judgment statements in the code. And it is a little difficult for me to find a proper place.

**Reply by jinningwang** [reply]
In the ``discrete`` module, there are ``LessThan`` and ``IsEqual`` can be used in your function. You can do a search and take a look at any models that have these elements.
